### DISCUSSION TITLE
Is it possible to have a Block Entity/Peripheral with different methods based on which block it's connected to?

### DISCUSSION THREAD

**Discussion by Povstalec** [cc-tweaked/CC-Tweaked#1379] [Q&A]
So, I want to have an Interface Block which can connect Computers to some of my Block Entities. 
Now, since a lot of those Block Entities share the same methods, with only a few unique methods for each one, it seems redundant to have a different version of the same Interface Block for each one. 

The first thing I thought of was simply making multiple peripherals extend from one with all the shared methods and then returning a different Peripheral through getCapability() based on which Block the Interface is connected to.
Though I immediately noticed a problem, that being the fact the Interface can't switch the Peripheral when I replace the Block Entity it's connected to. Which honestly isn't that much of a problem, since it would probably only ever be placed one time, but I'd still prefer to have the option for it to change on the fly.

And so I come to ask: Is there a way to decide which methods a Peripheral can use that can change over time? Or even simply a way to switch the Peripheral that's used when a Block Entity it's connected to changes?

**Comment by SquidDev** [comment]
So there's [`IDynamicPeripheral`](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.19.x/projects/core-api/src/main/java/dan200/computercraft/api/peripheral/IDynamicPeripheral.java), which allows you to dynamically define its methods. See for instance, [the implementation of `GenericPeripheral`](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.19.x/projects/common/src/main/java/dan200/computercraft/shared/peripheral/generic/GenericPeripheral.java).

That said, I wonder if you could just get away without the interface at all and just expose a peripheral directly on the targetted blocks?

**Reply by SquidDev** [reply]
Another alternative would be to invalidate and recreate the `LazyOptional` whenever the adjacent block changes, which causes he peripheral to be refetched.

**Reply by Povstalec** [reply]
> That said, I wonder if you could just get away without the interface at all and just expose a peripheral directly on the targetted blocks?

Technically I could, but it wouldn't fit the theme I'm going for with the Block Entity

> So there's [`IDynamicPeripheral`](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.19.x/projects/core-api/src/main/java/dan200/computercraft/api/peripheral/IDynamicPeripheral.java), which allows you to dynamically define its methods. See for instance, [the implementation of `GenericPeripheral`](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.19.x/projects/common/src/main/java/dan200/computercraft/shared/peripheral/generic/GenericPeripheral.java).

And thanks, I'm definitely going to play around with that, now that I know it exists!

**Reply by Povstalec** [reply]
> Another alternative would be to invalidate and recreate the `LazyOptional` whenever the adjacent block changes, which causes he peripheral to be refetched.

I have a question regarding this, as I've only been able to make it work one way. By that I mean: If I have the Block Entity there and run peripheral.wrap("back") the returned methods are correct, then if I break the Block Entity and wrap it again, the methods for that specific Block Entity are gone.
But then if I put the Block Entity back and wrap it one more time, the methods won't reappear.

I'm simply invalidating it and then returning a new LazyOptional in getCapability() is there something else I'm supposed to do besides that?

**Reply by SquidDev** [reply]
When you refer to "block entity" here, is this your interface next to the computer, or the block the interface is exposing methods for?

You may need to fire a block update if a peripheral becomes available.

**Reply by Povstalec** [reply]
I'm refering to the Block the Interface is exposing methods to.

As for the Block Updates, I've tried running setChanged() (I've even tried running it every tick at one point) and the outcome was the same.

EDIT: I don't know if this is important, but when printing out each step, I've noticed getCapability() is run 2 times, the first time with the old Block Values and second time with the new ones.

**Reply by SquidDev** [reply]
Oh, you want to be updating neighbours with `level.updateNeighborsAt` (or possibly `updateNeighborsAtExceptFromFacing`). `setChanged` just updates comparators (and marks the chunk as needing to be saved), while `updateNeighborsAt` fires block updates.

**Reply by Povstalec** [reply]
Oh, I did not know `setChanged` updated only comparators. Well, after using `updateNeighborsAt` there is still no difference.
(I've also changed it so that `getCapability()` is not run 2 times, but that didn't change anything at all)
